src/flexmock/api.py

```diff
@@ -392,77 +381,66 @@ def _handle_matched_expectation(
                 )
             expectation.times_called += 1
             expectation.verify(final=False)
-            _pass_thru = _getattr(expectation, "_pass_thru")
-            _replace_with = _getattr(expectation, "_replace_with")
-            if _pass_thru:
+            if expectation._pass_thru:
                 return pass_thru(expectation, runtime_self, *kargs, **kwargs)
-            if _replace_with:
-                return _replace_with(*kargs, **kwargs)
-            return_values = _getattr(expectation, "_return_values")
-            if return_values:
-                return_value = return_values[0]
-                del return_values[0]
-                return_values.append(return_value)
-            else:
-                return_value = ReturnValue()
-            if return_value.raises:
-                if inspect.isclass(return_value.raises):
-                    raise return_value.raises(
-                        *return_value.value["kargs"], **return_value.value["kwargs"]
-                    )
-                raise return_value.raises  # pylint: disable=raising-bad-type
-            return return_value.value
+            if expectation._replace_with:
+                return expectation._replace_with(*kargs, **kwargs)
+            return_values = expectation._return_values or [ReturnValue()]
+            return_value = return_values.pop(0)
+            return_values.append(return_value)
```

Comment:
> This is different from the existing behavior. If return_values is empty, here return_values would be [ReturnValue()] and in the existing code it is empty (ReturnValue() is never appended to return_values).

Actually not, because if `expectation._return_value` is empty, then the (mutable) reference to it is not what gets assigned to the `return_values` variable, so `expectation._return_value` would not be mutated at all in the case.

Reply:
Ah, right. Very confusing 😁 Anyway I think early return when `return_values`is empty would make this better.